### ISSUE TITLE
numpy.cov() and numpy.var() default bias is inconsistent (numpy 1.9.2)

### ISSUE THREAD

**Issue by jg-you** [numpy/numpy#5835]
cov() uses a bias of 1 by default.
var() uses a bias of 0 by default.

Such that
```python
import random
import numpy as np

x = [random.randint(0,1) for x in range(0,100)]
y = [random.randint(0,1) for x in range(0,100)]

if np.isclose(np.cov([x,x])[0,0], np.var(x)):
    print("Consistent by default.")

if np.isclose(np.cov([x,x],ddof=0)[0,0], np.var(x,ddof=0))
    print("Consistent.")
```

will only print the second line.

**Comment by charris** [comment]
Yes. I don't know what we can do about that at this point. This issue is probably best discussed on the mailing list.

**Comment by rgommers** [comment]
> ... of `cov`, which I guess is used less than `var`, so that's good.

That seems to be true (summary usage data from [here](https://raw.githubusercontent.com/data-apis/python-record-api/master/data/typing/numpy.py)):

<details>

```
def cov(
    m: object,
    y: object = ...,
    rowvar: Union[float, bool, int] = ...,
    bias: Union[float, int, bool] = ...,
    aweights: numpy.ndarray = ...,
):
    """
    usage.dask: 11
    usage.matplotlib: 3
    usage.pandas: 7
    usage.scipy: 21
    usage.sklearn: 24
    """

def var(
    a: object,
    axis: Union[int, None, Tuple[Union[int, None], ...]] = ...,
    out: Union[dask.dataframe.core.Scalar, dask.dataframe.core.Series] = ...,
    keepdims: bool = ...,
    dtype: Union[Literal["i8", "f8"], Type[float], None] = ...,
    ddof: int = ...,
):
    """
    usage.dask: 59
    usage.pandas: 13
    usage.scipy: 19
    usage.sklearn: 55
    usage.xarray: 31
    """
```

</details>

> That means changing the behavior of `cov`

Unfortunately I don't think we can do that. We could deprecate the function and add a new one (which would be painful), but we should definitely not change the behaviour of the current function - that would silently change numerical results and make currently valid code wrong. We try never to do that; a `FutureWarning` isn't enough to guarantee that people see the issue.

**Comment by AllenDowney** [comment]
@rgommers Yeah, "never break valid code" is a pretty good rule.

How about a warning if you call `cov` without `bias` or `ddof`, and then never change the behavior?

**Comment by rgommers** [comment]
> How about a warning if you call `cov` without `bias` or `ddof`, and then never change the behavior?

That does seem like a reasonable thing to do. Better than deprecating `cov`. Having to add `ddof=0/1` is a slight annoyance but makes the code more understandable, so I like the idea.

**Comment by icastorm** [comment]
> > How about a warning if you call `cov` without `bias` or `ddof`, and then never change the behavior?
> 
> That does seem like a reasonable thing to do. Better than deprecating `cov`. Having to add `ddof=0/1` is a slight annoyance but makes the code more understandable, so I like the idea.

Hello, I just wanted to add that this is something I ran into, and it cost me a day of work as I tried to trace the steps of what was going wrong in some code I had written. Someone smarter than me probably could've figured it out faster, but I figured maybe my experience as a less-than-genius user of numpy might be valuable.

The issue, of course, was that I didn't realize that numpy.cov() essentially uses a default of 1 for ddof. Not only was the behavior different from what I expected (which was the simple statistic, the same as numpy.var), the use of "None" as the real default implied to me at first glance that the default behavior **was** to do what I expected, which was ddof=0. Reading through the docs now the behavior does makes sense as written, but you have to interpret the behavior of the bias keyword as well to understand what is going on. Because I didn't look closely enough, I moved onto other things and it took me a while to circle back.

The changelog for this inconsistency goes all the way back to the early days of numpy, it looks like. The variance function was added first with ddof, and then covariance was added by a different developer using "bias", which is the convention in scipy as well. A third developer, seeking to unify the two, added the ddof argument to covariance but gave it a default of "None" so it didn't change the default behavior.

Because this is a problem that can't be solved by changing the behavior, I'd like to suggest a twofold solution to numpy.cov:


1. Change the default of ddof to 1
2. Include a warning when neither are specified

Because this doesn't change the default behavior or the hierarchy between the "ddof" and "bias" keywords, I don't think it would break any existing code and would make the default behavior a lot more clear to someone like me. I actually mentioned this to several friends at work who were horrified and found similar bugs in their own code because they didn't specify ddof. While the warnings would be ugly, it might just save a lot of users from headaches down the line.
